### DISCUSSION TITLE
KMP Pagination

### DISCUSSION THREAD

**Discussion by blvckcoded** [AndroidDev-social/DodoForMastodon#52] [General]
We will need to support pagination for the timeline(s). 

Some quick R&D and I found these two:
https://github.com/kuuuurt/multiplatform-paging
https://github.com/cashapp/multiplatform-paging

We need to confirm that desktop is also supported, at first glance, I just see iOS and Android. 

[paging-common in cashapp](https://github.com/cashapp/multiplatform-paging/blob/main/paging-common/build.gradle.kts#L11-L16) shows jvm as a target, but that is not the case with [paging-runtime](https://github.com/cashapp/multiplatform-paging/blob/main/paging-runtime/build.gradle.kts#L10-L13)

Just creating this to get us started

**Comment by kpgalligan** [comment]
Brain-dumping some other research. `paging-common` on the JVM typealiases `paging-common` in the Android source, which appears to target the general JVM. `paging-runtime` in Android source has Android-specific code in it. `paging-runtime` in the Cash app library has an [iOS implementation](https://github.com/cashapp/multiplatform-paging/blob/main/paging-runtime/src/iosMain/kotlin/app/cash/paging/PagingCollectionViewController.kt), which doesn't look too big, so the scope of what would likely need to be implemented seems reasonable.

I'll preface the next statement by saying I've never used paging the library, and my research was kind of a quick overview. However...

There's `paging-runtime` for Android, which has some "classic" Android UI in it. There's also `paging-compose` [here](https://android.googlesource.com/platform/frameworks/support/+/HEAD/paging/paging-compose). If that is what you'd need for a Compose UI app on Android, looking at the dependencies:

```groovy
    api("androidx.compose.foundation:foundation:1.0.5")
    api(project(":paging:paging-common"))
    api("androidx.compose.runtime:runtime:1.0.5")
```

That should be *copyable* to Compose Desktop running on the JVM without much trouble. I doubt just using that dependency would work, but it might? The actual code itself, though, should work without much change, if any.

In theory that should *also* work for Compose UI on iOS, but that is very far from stable. UI on iOS is a whole different discussion.

Anyway, assuming the `paging-compose` code itself works for Desktop, there's room to contribute that support back to the Cash app library, or at least publish something in tandem if anybody else wants to use it.

**Comment by veyndan** [comment]
> [paging-common in cashapp](https://github.com/cashapp/multiplatform-paging/blob/main/paging-common/build.gradle.kts#L11-L16) shows jvm as a target, but that is not the case with [paging-runtime](https://github.com/cashapp/multiplatform-paging/blob/main/paging-runtime/build.gradle.kts#L10-L13)

I think it's a bit confusing looking at the source code, but there is Android support (just not from us!).

As `paging-common` in Multiplatform Paging type aliases to `paging-common` in AndroidX Paging, you can just use the `paging-runtime` supplied by AndroidX Paging on Android. Here's [the section discussing this](https://github.com/cashapp/multiplatform-paging/#interoperability-with-androidx-paging) in the README, but a high level dependency overview for an Android and iOS app is:

```
src/
├─ commonMain/
│  ├─ implementation("app.cash.paging:paging-common")
├─ androidMain/
│  ├─ implementation("androidx.paging:paging-runtime") OR implementation("androidx.paging:paging-compose")
├─ iosMain/
│  ├─ implementation("app.cash.paging:paging-runtime")
```

> We need to confirm that desktop is also supported, at first glance, I just see iOS and Android.

At the moment, we only support Android and iOS. We're definitely open to contributions to add desktop support!

To add support, you'll need to:

* Add support for the necessary targets in `paging-common`. This _should_ be fairly straightforward, as all the code is in `commonMain`. I'd be happy to help out with this part if you face any issues. You can skip this step if you just want to support JVM desktop, as the JVM is already a target.
* Add the desktop targets to `paging-runtime`. This is basically providing a list view that can display pages provided by `paging-common`. I'm still happy to help with this part, but I have no experience with developing desktop apps so don't know how much help I'll be. Checking out [the iOS implementation of `paging-runtime`](https://github.com/cashapp/multiplatform-paging/blob/main/paging-runtime/src/iosMain/kotlin/app/cash/paging/PagingCollectionViewController.kt), the main thing that the desktop list view needs to handle is notifications of item insertion, removal, moves, and changes.

**Comment by kpgalligan** [comment]
> I think it's a bit confusing looking at the source code, but there is Android support (just not from us!).

Right.

> Add the desktop targets to paging-runtime.

As stated, if `paging-compose` from the Jetpack Paging library is the Compose UI equivalent of `paging-runtime` for the classing Android UI, then I don't think you really need to implement much of a new "thing". You can just use that source in the Desktop version because they're both (basically) stock Compose UI. Also, in theory, you could use that for Compose UI for iOS, although I don't think you'd want to publish a serious UI with Compose UI on iOS right now.

**Reply by veyndan** [reply]
Yep, I agree with your approach, and that Compose UI is probably the easiest way to go for Multiplatform Paging support. The original comment was in response to Omid, to give a general impression of what adding desktop support to Multiplatform Paging would look like, regardless of the framework.

**Reply by kpgalligan** [reply]
I just wanted to make extra clear that I think this could be almost free, before somebody started hacking away at writing a new implementation. Also, I guess it should be pointed out, the Android client probably wouldn't use `paging-runtime` either, but `paging-compose` instead, but I haven't actually used `paging-compose`. I should probably just try it...

**Reply by veyndan** [reply]
I've used `paging-compose` on Android, but have never done any desktop Compose UI stuff, so can't comment on that. However, having looked at the source of `paging-compose` before (which only consists of the file [`LazyPagingItems.kt`](https://github.com/androidx/androidx/blob/androidx-main/paging/paging-compose/src/main/java/androidx/paging/compose/LazyPagingItems.kt)), the only part that is Android specific are [these lines](https://github.com/androidx/androidx/blob/6571b15822a6371389176ff9f56bf213ea2f627e/paging/paging-compose/src/main/java/androidx/paging/compose/LazyPagingItems.kt#L283-L381), which are basic UI bindings to the rest of the code. The rest of the file just uses Compose and could be relocated to `commonMain` and shared across platforms.

**Comment by blvckcoded** [comment]
I opened up a draft PR against the cashapp library. If anyone is interested, please feel free to follow along there https://github.com/cashapp/multiplatform-paging/pull/9
